File: examples/tinybc/tinybc.cpp
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include "sgl/sgl.h"
#include "sgl/core/platform.h"
#include "sgl/core/bitmap.h"
#include "sgl/core/timer.h"
#include "sgl/device/device.h"
#include "sgl/device/shader.h"
#include "sgl/device/command.h"
#include "sgl/device/shader_cursor.h"
#include "sgl/device/shader_object.h"
#include "sgl/device/kernel.h"
#include "sgl/device/query.h"
#include "sgl/device/agility_sdk.h"
#include "sgl/utils/tev.h"

#include <argparse/argparse.hpp>

SGL_EXPORT_AGILITY_SDK

static const std::filesystem::path EXAMPLE_DIR(SGL_EXAMPLE_DIR);

using namespace sgl;

int main(int argc, const char* argv[])
{
    argparse::ArgumentParser args("tinybc");
    args.add_description("Slang-based BC7 - mode 6 compressor");
    args.add_argument("input_path").required().help("Path to the input texture.");
    args.add_argument("-o", "--output_path").help("Optional path to save the decoded BC7 texture.");
    args.add_argument("-s", "--opt_steps").default_value(100).help("Number of optimization (gradient descene) steps.");
    args.add_argument("-b", "--benchmark")
        .default_value(false)
        .implicit_value(true)
        .help("Run in benchmark mode to measure processing time.");
    args.add_argument("-t", "--tev").default_value(false).implicit_value(true).help("Show images in tev image viewer.");
    args.add_argument("-v", "--verbose").default_value(false).implicit_value(true).help("Enable verbose logging.");

    try {
        // args.parse_args(argc, argv);
        SGL_UNUSED(argc, argv);
        args.parse_args(
            {"tinybc",
             "C:/src/sgl/data/test_images/monalisa.jpg",
             "-o",
             "C:/src/sgl/monalisa_bc7.jpg",
             "-t",
             "-b",
             "-v"}
        );
    } catch (const std::runtime_error& err) {
        std::cerr << err.what() << std::endl;
        return 1;
    }

    std::string input_path = args.get("input_path");
    std::optional<std::string> output_path = args.present("output_path");
    int opt_steps = args.get<int>("opt_steps");
    bool benchmark = args.get<bool>("benchmark");
    bool tev = args.get<bool>("tev");
    bool verbose = args.get<bool>("verbose");

    sgl::static_init();

    {
        ref<Bitmap> input;

        try {
            input = make_ref<Bitmap>(input_path)
                        ->convert(Bitmap::PixelFormat::rgba, Bitmap::ComponentType::float32, false);
        } catch (const std::exception& e) {
            fmt::println("Failed to load input image from {}:\n{}", input_path, e.what());
            return 1;
        }

        uint32_t w = input->width();
        uint32_t h = input->height();

        ref<Device> device = Device::create({
            .enable_debug_layers = verbose,
            .compiler_options = {.include_paths = {EXAMPLE_DIR}},
        });

        SubresourceData initial_data[1] = {{
            .data = input->data(),
            .size = input->buffer_size(),
            .row_pitch = input->width() * 4 * sizeof(float),
        }};

        // Create input texture
        ref<Texture> input_tex = device->create_texture({
            .format = Format::rgba32_float,
            .width = w,
            .height = h,
            .usage = TextureUsage::shader_resource,
            .data = initial_data,
        });

        // Show input texture in tev
        if (tev)
            tev::show_async(input_tex, "tinybc-input");

        // Create decoded texture
        ref<Texture> decoded_tex = device->create_texture({
            .format = Format::rgba32_float,
            .width = w,
            .height = h,
            .usage = TextureUsage::unordered_access,
        });

        std::string constants = fmt::format(
            "export static const bool USE_ADAM = true;\n"
            "export static const uint OPT_STEPS = {};\n",
            opt_steps
        );
        ref<ShaderProgram> program = device->load_program("tinybc.slang", {"main"}, constants);
        ref<ComputeKernel> kernel = device->create_compute_kernel({.program = program});

        uint32_t num_iters = benchmark ? 1000 : 1;

        Timer t;

        // Setup query pool to measure GPU time
        ref<QueryPool> queries = device->create_query_pool({.type = QueryType::timestamp, .count = num_iters * 2});

        // Compress!
        ref<CommandEncoder> command_encoder = device->create_command_encoder();
        for (uint32_t i = 0; i < num_iters; ++i) {
            command_encoder->write_timestamp(queries, i * 2);
            kernel->dispatch(
                uint3(w, h, 1),
                [&](ShaderCursor cursor)
                {
                    cursor["input_tex"] = input_tex;
                    cursor["decoded_tex"] = decoded_tex;
                    cursor["lr"] = 0.1f;
                    cursor["adam_beta_1"] = 0.9f;
                    cursor["adam_beta_2"] = 0.999f;
                },
                command_encoder
            );
            command_encoder->write_timestamp(queries, i * 2 + 1);
        }
        device->submit_command_buffer(command_encoder->finish());

        device->wait();

        double total_cpu_time_sec = t.elapsed_s();

        std::vector<double> times = queries->get_timestamp_results(0, num_iters * 2);
        double comp_time_sec = 0.0;
        for (uint32_t i = 0; i < num_iters; ++i)
            comp_time_sec += (times[i * 2 + 1] - times[i * 2]);
        comp_time_sec /= num_iters;

        // Calculate and print performance metrics
        if (benchmark) {
            double textures_per_sec = 1.0 / comp_time_sec;
            double giga_texels_per_sec = w * h * textures_per_sec / 1e9;
            fmt::println("Benchmark:");
            fmt::println("- Number of optimization steps: {}", opt_steps);
            fmt::println("- Compression time: {:.4g} ms", comp_time_sec * 1e3);
            fmt::println("- Compression throughput: {:.4g} GTexels/s", giga_texels_per_sec);
            fmt::println("- Total CPU time: {:.4g} s", total_cpu_time_sec);
        }

        // Calculate and print PSNR
        ref<Bitmap> decoded = decoded_tex->to_bitmap();
        double mse = 0.0;
        const float* input_data = input->data_as<float>();
        const float* decoded_data = decoded->data_as<float>();
        for (uint32_t i = 0; i < w * h * 4; ++i)
            mse += (input_data[i] - decoded_data[i]) * (input_data[i] - decoded_data[i]);
        mse /= w * h * 4;
        double psnr = 20.0 * std::log10(1.0 / std::sqrt(mse));
        fmt::println("PSNR: {:.4g}", psnr);

        // Show decoded texture in tev
        if (tev)
            tev::show_async(decoded_tex, "tinybc-decoded");

        // Output decoded texture
        if (output_path) {
            decoded_tex->to_bitmap()
                ->convert(Bitmap::PixelFormat::rgb, Bitmap::ComponentType::uint8, true)
                ->write_async(*output_path);
        }

        device->close();
    }

    sgl::static_shutdown();

    return 0;
}

File: examples/tinybc/tinybc.py
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

import sys
import argparse
import numpy as np
import slangpy as spy
from pathlib import Path

EXAMPLE_DIR = Path(__file__).parent

# Build and parse command line
# fmt: off
parser = argparse.ArgumentParser(description="Slang-based BC7 - mode 6 compressor")
parser.add_argument("input_path", help="Path to the input texture.")
parser.add_argument("-o", "--output_path", help="Optional path to save the decoded BC7 texture.")
parser.add_argument("-s", "--opt_steps", type=int, default=100, help="Number of optimization (gradient descene) steps.")
parser.add_argument("-b", "--benchmark", action="store_true", help="Run in benchmark mode to measure processing time.")
parser.add_argument("-t", "--tev", action="store_true", help="Show images in tev image viewer.")
parser.add_argument("-v", "--verbose", action="store_true", help="Enable verbose logging.")
# fmt: on
# args = parser.parse_args()
args = parser.parse_args(
    [
        str(EXAMPLE_DIR.parent.parent / "data/test_images/monalisa.jpg"),
        "-o",
        "monalisa_bc7.jpg",
        "-t",
        "-b",
        "-v",
    ]
)

# Load input image
try:
    image = spy.Bitmap(args.input_path).convert(
        pixel_format=spy.Bitmap.PixelFormat.rgba,
        component_type=spy.Bitmap.ComponentType.float32,
        srgb_gamma=False,
    )
    w, h = image.width, image.height
    input = np.asarray(image)
except Exception as e:
    print(f"Failed to load input image from {args.input_path}:\n{e}")
    sys.exit(1)

# Create device
device = spy.Device(
    enable_debug_layers=args.verbose,
    compiler_options={"include_paths": [EXAMPLE_DIR]},
)

# Create input texture
input_tex = device.create_texture(
    format=spy.Format.rgba32_float,
    width=w,
    height=h,
    usage=spy.TextureUsage.shader_resource,
    data=input,
)

# Show input texture in tev
if args.tev:
    spy.tev.show_async(input_tex, name="tinybc-input")

# Create decoded texture
decoded_tex = device.create_texture(
    format=spy.Format.rgba32_float,
    width=w,
    height=h,
    usage=spy.TextureUsage.unordered_access,
)

# Load shader module
constants = f"export static const bool USE_ADAM = true;\nexport static const uint OPT_STEPS = {args.opt_steps};\n"
program = device.load_program("tinybc.slang", ["compute_main"], constants)
kernel = device.create_compute_kernel(program)

t = spy.Timer()

# When running in benchmark mode amortize overheads over many runs to measure more accurate GPU times
num_iters = 1000 if args.benchmark else 1

# Setup query pool to measure GPU time
queries = device.create_query_pool(spy.QueryType.timestamp, num_iters * 2)

# Compress!
command_encoder = device.create_command_encoder()
for i in range(num_iters):
    command_encoder.write_timestamp(queries, i * 2)
    kernel.dispatch(
        thread_count=[w, h, 1],
        vars={
            "input_tex": input_tex,
            "decoded_tex": decoded_tex,
            "lr": 0.1,
            "adam_beta_1": 0.9,
            "adam_beta_2": 0.999,
        },
        command_encoder=command_encoder,
    )
    command_encoder.write_timestamp(queries, i * 2 + 1)
device.submit_command_buffer(command_encoder.finish())

# Wait for GPU to finish and get timestamps
device.wait()

total_cpu_time_sec = t.elapsed_s()

times = np.asarray(queries.get_timestamp_results(0, num_iters * 2))
comp_time_sec = np.mean(times[1::2] - times[0::2])

# Calculate and print performance metrics
if args.benchmark:
    textures_per_sec = 1 / comp_time_sec
    giga_texels_per_sec = w * h * textures_per_sec / 1e9
    print(f"Benchmark:")
    print(f"- Number of optimization steps: {args.opt_steps}")
    print(f"- Compression time: {1e3 * comp_time_sec:.4g} ms")
    print(f"- Compression throughput: {giga_texels_per_sec:.4g} GTexels/s")
    print(f"- Total CPU time: {total_cpu_time_sec:.4g} s")

# Calculate and print PSNR
decoded = decoded_tex.to_numpy()
mse = np.mean((input - decoded) ** 2)
psnr = 20 * np.log10(1.0 / np.sqrt(mse))
print(f"PSNR: {psnr:.4g}")

# Show decoded texture in tev
if args.tev:
    spy.tev.show_async(decoded_tex, name="tinybc-decoded")

# Output decoded texture
if args.output_path:
    decoded_tex.to_bitmap().convert(
        pixel_format=spy.Bitmap.PixelFormat.rgb,
        component_type=spy.Bitmap.ComponentType.uint8,
        srgb_gamma=True,
    ).write_async(args.output_path)
